src/Workspaces/Core/Portable/Options/GlobalOptionService.cs

```diff
@@ -236,27 +243,103 @@ public bool RefreshOption(OptionKey2 optionKey, object? newValue)
             return true;
         }
 
+        public void AddOptionChangedHandler(object target, EventHandler<OptionChangedEventArgs> handler)
+        {
+            lock (_gate)
+            {
+                if (_keepAliveTable.TryGetValue(target, out var existingHandler))
+                {
+                    var newHandler = existingHandler + handler;
+#if NET6_0_OR_GREATER
+                    _keepAliveTable.AddOrUpdate(target, newHandler);
+#else
+                    _keepAliveTable.Remove(target);
+                    _keepAliveTable.Add(target, newHandler);
+#endif
+                }
+            }
+
+            ImmutableInterlocked.Update(
+                ref _weakHandlers,
+                static (weakHandlers, handler) =>
+                {
+                    return weakHandlers
+                        .RemoveAll(WeakReferenceExtensions.IsNull)
+                        .Add(new WeakReference<EventHandler<OptionChangedEventArgs>>(handler));
+                },
+                handler);
+        }
+
+        public void RemoveOptionChangedHandler(object target, EventHandler<OptionChangedEventArgs> handler)
+        {
+            lock (_gate)
+            {
+                if (_keepAliveTable.TryGetValue(target, out var existingHandler))
+                {
+                    var newHandler = existingHandler - handler;
+                    if (newHandler is null)
+                    {
+                        _keepAliveTable.Remove(target);
+                    }
+                    else
+                    {
+#if NET6_0_OR_GREATER
+                        _keepAliveTable.AddOrUpdate(target, newHandler);
+#else
+                        _keepAliveTable.Remove(target);
+                        _keepAliveTable.Add(target, newHandler);
+#endif
+                    }
+                }
+            }
+
+            ImmutableInterlocked.Update(
+                ref _weakHandlers,
+                static (weakHandlers, handler) =>
+                {
+                    return weakHandlers
+                        .RemoveAll(weakHandler => !weakHandler.TryGetTarget(out var target) || (target - handler) is null);
```

Comment:
this seems odd too.  so we have the existing target, and we remove hte handler from it (getting a new value back), but if there are existing handlers... we don't do anything.  but the actual target wasn't updated.  it was just `-` and not `-=` (morally).

Reply:
➡️ Expanded and simplified this logic in `WeakEvent<TEventArgs>`